lib/csv.rb

```diff
@@ -1590,14 +1600,19 @@ def open(filename, mode="r", **options)
       options.delete(:replace)
       options.delete_if {|k, _| /newline\z/.match?(k)}
 
-      begin
-        f = File.open(filename, mode, **file_opts)
-      rescue ArgumentError => e
-        raise unless /needs binmode/.match?(e.message) and mode == "r"
-        mode = "rb"
-        file_opts = {encoding: Encoding.default_external}.merge(file_opts)
-        retry
+      if filename_or_io.is_a?(StringIO)
+        f = filename_or_io
```

Comment:
How about creating a new `StringIO` instead of reusing the given `StringIO` like we did for `IO`?

Reply:
Good point, I missed that the `File.open` call to an already open file stream creates a new IO.